ext/ddtrace_profiling_native_extension/profiling.c

```diff
@@ -59,6 +66,70 @@ static VALUE native_working_p(DDTRACE_UNUSED VALUE _self) {
   return Qtrue;
 }
 
+struct trigger_grab_gvl_and_raise_arguments {
+  VALUE exception_class;
+  char *test_message;
+  int test_message_arg;
+};
+
+static VALUE _native_grab_gvl_and_raise(DDTRACE_UNUSED VALUE _self, VALUE exception_class, VALUE test_message, VALUE test_message_arg) {
+  ENFORCE_TYPE(test_message, T_STRING);
+
+  struct trigger_grab_gvl_and_raise_arguments args;
+
+  args.exception_class = exception_class;
+  args.test_message = StringValueCStr(test_message);
+  args.test_message_arg = test_message_arg != Qnil ? NUM2INT(test_message_arg) : -1;
+
+  rb_thread_call_without_gvl(trigger_grab_gvl_and_raise, &args, NULL, NULL);
+
+  rb_raise(rb_eRuntimeError, "This should never happen");
```

Comment:
If I'm reading this right and this should be unreachable, unless we messed up code in `trigger_grab_gvl_and_raise`,  I think the error could be more descriptive, similar to `  rb_bug("[DDTRACE] Unexpected: Reached the end of grab_gvl_and_raise while raising '%s'\n", args.exception_message);`.

Reply:
Since this is only for testing the code, do you think it's worth it?

Because this `_native_grab_gvl_and_raise` is directly called from Ruby code, the stack trace seems pretty self-explanatory as well:

```
  3) Datadog::Profiling::NativeExtension grab_gvl_and_raise_syserr raises an exception with the passed in message and errno
     Failure/Error:
       expect do
         described_class::Testing._native_grab_gvl_and_raise_syserr(Errno::EINTR::Errno, 'this is a test', nil)
       end.to raise_exception(Errno::EINTR, "#{Errno::EINTR.exception.message} - this is a test")

       expected Errno::EINTR with "Interrupted system call - this is a test", got #<RuntimeError: This should never happen> with backtrace:
         # ./spec/datadog/profiling/native_extension_spec.rb:186:in `_native_grab_gvl_and_raise_syserr'
         # ./spec/datadog/profiling/native_extension_spec.rb:186:in `block (4 levels) in <top (required)>'
         # ./spec/datadog/profiling/native_extension_spec.rb:185:in `block (3 levels) in <top (required)>'
         # ./spec/spec_helper.rb:216:in `block (2 levels) in <top (required)>'

#...

  6) Datadog::Profiling::NativeExtension grab_gvl_and_raise raises the requested exception with the passed in message
     Failure/Error:
       expect { described_class::Testing._native_grab_gvl_and_raise(ZeroDivisionError, 'this is a test', nil) }
         .to raise_exception(ZeroDivisionError, 'this is a test')

       expected ZeroDivisionError with "this is a test", got #<RuntimeError: This should never happen> with backtrace:
         # ./spec/datadog/profiling/native_extension_spec.rb:166:in `_native_grab_gvl_and_raise'
         # ./spec/datadog/profiling/native_extension_spec.rb:166:in `block (4 levels) in <top (required)>'
         # ./spec/datadog/profiling/native_extension_spec.rb:166:in `block (3 levels) in <top (required)>'
         # ./spec/spec_helper.rb:216:in `block (2 levels) in <top (required)>'
```

(Here I simulated the issue by commenting out the `rb_thread_call_without_gvl` lines on both functions).

Reply:
Ah, I see your point -- indeed the exception is written from the POV of someone reading the source, but if you just get it somehow, it's not clear what "this" is. Tweaked the messages in 831bf46de688f2853c636469957ab5c1b40ef044 .